core/src/main/c/share/bitmap_index_utils.cpp

```diff
@@ -0,0 +1,153 @@
+            continue;
+        }
+
+        if (rows[k] > 0) continue;
+
+        int64_t value_count = 0;
+        int64_t last_vblock_offset = 0;
+        int64_t first_vblock_offset = 0;
+        auto retries_count = 10;
+        bool is_kblock_consistent = false;
+        while(retries_count--) {
+            value_count = keys[k].value_count;
+            std::atomic_thread_fence(std::memory_order_acquire);
+            if (keys[k].count_check == value_count) {
+                first_vblock_offset = keys[k].first_value_block_offset;
+                last_vblock_offset = keys[k].last_value_block_offset;
+                std::atomic_thread_fence(std::memory_order_acquire);
+                if (keys[k].value_count == value_count) {
+                    is_kblock_consistent = true;
+                    break;
+                }
+            }
+        }
+        bool update_range = true;
+        if(value_count > 0) {
+            int64_t vblock_end_offset = last_vblock_offset + vblock_size;
+            bool is_offset_in_mapped_area = vblock_end_offset <= value_memory_size;
+
+            if (!is_kblock_consistent || !is_offset_in_mapped_area) {
+                // can trust only first vblock offset
+                last_vblock_offset = first_vblock_offset;
+                auto link = reinterpret_cast<const value_block_link *>(values_memory + last_vblock_offset + vblock_size - sizeof(value_block_link));
+                int64_t block_traversed = 1;
+                while(link->next && link->next + vblock_size <= value_memory_size) {
+                    last_vblock_offset = link->next;
+                    link = reinterpret_cast<const value_block_link *>(values_memory + last_vblock_offset + vblock_size - sizeof(value_block_link));
+                    block_traversed += 1;
+                }
+                //assuming blocks are full
+                value_count = vblock_capacity*block_traversed;
+            }
+
+            auto values = reinterpret_cast<const int64_t *>(values_memory);
+            auto res = seekValueBlockRTL(value_count, last_vblock_offset, values, max_value,
+                                         vblock_capacity_mask);
+            value_count = res.first;
+            last_vblock_offset = res.second;
+            if (value_count > 0) {
+                uint64_t cell_index = --value_count & vblock_capacity_mask;
+                int64_t local_row_id = values[last_vblock_offset / 8 + cell_index];
+                if (local_row_id >= min_value) {
+                    rows[k] = ((int64_t) partition_index << 44) + local_row_id + 1;
+                    row_count += 1;
+                    update_range = false;
+                }
+            }
+        }
+
+        // unindexed nulls case
+        if (k == 0 && unindexed_null_count > 0) {
+            if (rows[k] <= 0 && unindexed_null_count - 1 >= min_value) {
+                rows[k] = ((int64_t) partition_index << 44) + unindexed_null_count;
```

Comment:
we should perhaps refactor a helper method here, which creates rowid, perhaps an abstract type ?